src/rgw/services/svc_bi_rados.cc

```diff
@@ -351,7 +367,8 @@ int RGWSI_BucketIndex_RADOS::clean_index(RGWBucketInfo& bucket_info)
   dir_oid.append(bucket_info.bucket.bucket_id);
 
   std::map<int, std::string> bucket_objs;
-  get_bucket_index_objects(dir_oid, bucket_info.layout.current_index.layout.normal.num_shards, &bucket_objs);
+  get_bucket_index_objects(dir_oid, bucket_info.layout.current_index.layout.normal.num_shards,
+                          gen, &bucket_objs);
```

Comment:
this isn't right, because it mixes the 'current' layout/num_shards with the previous layout's generation. the previous generation's shard count will be different

i still think we need to pass in a whole bucket_index_layout_generation. you're right that we can't pass `bucket_info.layout.current_index` directly, but we should make a copy of the old layout before do_reshard() overwrites it. ie:
```diff
   //overwrite current_index for the next reshard process
+  const auto prev_index = bucket_info.layout.current_index;
   bucket_info.layout.current_index = *bucket_info.layout.target_index;
...
-  int ret2 = store->svc()->bi->clean_index(bucket_info, bucket_info.layout.target_index->gen);
+  int ret2 = store->svc()->bi->clean_index(bucket_info, prev_index);
```

Reply:
> this isn't right, because it mixes the 'current' layout/num_shards with the previous layout's generation. the previous generation's shard count will be different

Yes you're right, I missed that. Thanks.

Reply:
done

Reply:
@cbodley also added some code to address recovery upon resharding failure in commit e6b60da.